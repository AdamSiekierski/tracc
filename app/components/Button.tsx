import React, { DetailedHTMLProps, HTMLAttributes, ReactNode } from "react"
import { IconType } from "./icons"

type ButtonProps = {
  children?: ReactNode
  variant?: "full" | "outlined"
  icon?: IconType
} & DetailedHTMLProps<HTMLAttributes<HTMLButtonElement>, HTMLButtonElement>

const Button = ({ children, variant, icon: Icon, ...props }: ButtonProps) => {
  let style: string
  switch (variant) {
    case "full":
      style = "bg-gray-200 hover:bg-gray-300"
      break
    case "outlined":
      style = "bg-gray-50 hover:bg-gray-100 border border-gray-200"
      break
    default:
      style = "bg-gray-200 hover:bg-gray-300"
      break
  }

  return (
    <button
      className={`block rounded-md px-4 py-2 text-gray-600 transition-colors duration-300 ${style}`}
      {...props}
    >
      {Icon && <Icon className="w-8 text-gray-300" />}
      {children}
    </button>
  )
}

export default Button
