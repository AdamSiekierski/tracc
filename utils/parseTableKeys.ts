export function parseTableKeys<T>(data: T[]) {
  return [
    ...data.reduce((acc, obj) => {
      for (const key of Object.keys(obj)) {
        acc.add(key)
      }

      return acc
    }, new Set<string>([])),
  ]
}
