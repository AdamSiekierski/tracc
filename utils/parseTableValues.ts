export function parseTableValues<T>(data: T[], keys: string[]): string[][] {
  return data.reduce((acc: string[][], obj: T) => {
    const currentValues: string[] = []

    for (const key of keys) {
      currentValues.push(obj[key] || "")
    }

    return [...acc, currentValues]
  }, [])
}
