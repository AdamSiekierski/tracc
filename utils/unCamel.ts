export const unCamel = (str) => str.replace(/([A-Z])/g, " $1").replace(/^./, (s) => s.toUpperCase())
