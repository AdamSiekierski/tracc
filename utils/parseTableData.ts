import { parseTableKeys } from "./parseTableKeys"
import { parseTableValues } from "./parseTableValues"

export function parseTableData<T>(data: T[]): [string[], string[][]] {
  const keys = parseTableKeys<T>(data)
  const rows = parseTableValues<T>(data, keys)

  return [keys, rows]
}
