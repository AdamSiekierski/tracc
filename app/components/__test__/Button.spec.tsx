import React from "react"
import Button from "../Button"
import { cleanup, fireEvent } from "@testing-library/react"
import { render } from "test/utils"

describe("Button", () => {
  const MockIcon = (props) => (
    <div {...props} data-testid="mock-icon">
      icon
    </div>
  )

  afterEach(cleanup)

  it("renders properly", () => {
    const wrapper = render(<Button data-testid="button">hello world</Button>)

    expect(wrapper.getByTestId("button")).toHaveTextContent("hello world")
    expect(wrapper.asFragment()).toMatchSnapshot()
  })

  it("calls onClick when clicked", () => {
    const onClick = jest.fn()

    const wrapper = render(<Button data-testid="button" onClick={onClick} />)

    fireEvent.click(wrapper.getByTestId("button"))

    expect(onClick).toHaveBeenCalledTimes(1)
    expect(onClick).toHaveBeenCalled()
  })

  it("shows the icon if passed", () => {
    const { getByTestId, asFragment } = render(<Button icon={MockIcon}>hello world</Button>)

    expect(getByTestId("mock-icon")).toBeInTheDocument()
    expect(getByTestId("mock-icon")).toHaveTextContent("icon")
    expect(asFragment()).toMatchSnapshot()
  })
})
