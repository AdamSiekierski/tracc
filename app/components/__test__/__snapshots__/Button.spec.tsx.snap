// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders properly 1`] = `
<DocumentFragment>
  <button
    class="block rounded-md px-4 py-2 text-gray-600 transition-colors duration-300 bg-gray-200 hover:bg-gray-300"
    data-testid="button"
  >
    hello world
  </button>
</DocumentFragment>
`;

exports[`Button shows the icon if passed 1`] = `
<DocumentFragment>
  <button
    class="block rounded-md px-4 py-2 text-gray-600 transition-colors duration-300 bg-gray-200 hover:bg-gray-300"
  >
    <div
      class="w-8 text-gray-300"
      data-testid="mock-icon"
    >
      icon
    </div>
    hello world
  </button>
</DocumentFragment>
`;
